tests/system/__init__.py

```diff
@@ -0,0 +1,176 @@
+# ^  For the sake of simplifying tests, we assume that even in Windows,
+#    env will be available (via msys/mingw)
+
+
+def is_venv():
+    """Check if the tests are running inside a venv"""
+    return hasattr(sys, 'real_prefix') or (sys.base_prefix != sys.prefix)
+
+
+@contextmanager
+def env(env_dict):
+    original = {k: environ[k] for k in env_dict.keys() if k in environ}
+    changed = [environ.putenv(k, v) or k for k, v in env_dict.items()]
+    try:
+        yield
+    finally:
+        # Restore original environment
+        for k in changed:
+            environ.unsetenv(k)
+        for k, v in original.items():
+            environ[k] = v
+
+
+def normalize_run_args(args):
+    """Make sure we have a flatten list of shell-words"""
+    if len(args) == 1:
+        if isinstance(args[0], str):
+            args = shlex.split(args[0], posix=not(IS_WIN))
+        elif isinstance(args[0], (list, tuple)):
+            args = args[0]
+
+    return list(args)
+
+
+def _insert_env(args):
+    """It seems that :mod:`subprocess` doesn't consider changes in the
+    environment variables after the main python process is started.
+    To workaround this we can always prepend the command with `env`, this
+    guarantees environment variables have up-to-date values.
+    """
+    if len(args) > 0 and args[0] not in ('env', ENV):
+        args.insert(0, ENV)
+
+    return args
+
+
+def run(*args, **kwargs):
+    """Run an external command (subprocess)
+
+    Arguments
+       args(str or list): list or string(s) representing the command::
+
+            >>> run('ls -la')
+            >>> run('ls', '-la')
+            >>> run(['ls', '-la'])
+
+    Keyword Arguments
+        env_vars(dict): Environment variables to set while running the command
+        **: Any other keyword argument from :func:`subprocess.run`
+
+    Returns
+        str: Everything sent to the stdout and stderr by the command
+    """
+    args = [str(a) for a in _insert_env(normalize_run_args(args))]
+    opts = dict(
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        universal_newlines=True,
+    )
+    opts.update(kwargs)
+    verbose = opts.pop('verbose', False)
+
+    with env(opts.pop('env_vars', {})):
+        if verbose:
+            # Print command and stdout in 2 steps to provide as much
+            # information as possible before an exception
+            print(*args, file=sys.stderr)
+        process = subprocess.run(args, **opts)
+        stdout = process.stdout.strip()
+        if verbose:
+            print(stdout, file=sys.stderr)
+        process.check_returncode()
+        return stdout
+
+
+class Venv(VirtualEnv):
+    """High-level interface for interacting with :mod:`venv` module.
+
+    Arguments
+        path(os.PathLike): path where the venv should be created
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        for k, v in environ.items():
+            if 'PIP_' in k:
```

Comment:
Hopefully this will allow us to use PIP_TRUSTED_HOST at least once, when updating `pip/setuptools/certifi`.
From there on, `setuptools` should be able to use `certifi` up-to-date certificates when installing build dependencies listed in `setup_requires` (Related to SSL errors on Windows, described in #244)